### DISCUSSION TITLE
How to detect query change using next/navigation in shallow mode

### DISCUSSION THREAD

**Discussion by MantasJa** [47ng/nuqs#558] [Q&A] [answered]
Hello guys, I would like to ask you how to detect query change using shallow routing and using next/navigation. This example does not work:

```
'use client'
 
import { usePathname, useSearchParams } from 'next/navigation'
 
function ExampleClientComponent() {
  const pathname = usePathname()
  const searchParams = useSearchParams()
  useEffect(() => {
    // Do something here...
  }, [pathname, searchParams])
}
```
Next.js 14.0.1

Why do I need this? I have few input components used in a search component, and I would like to initiate a server action  for updating search results in other component. But at the moment I can not detect route changes.

**Comment by franky47** [comment, marked as answer]
What version of `next` are you using? The `useSearchParams` hook only started to become reactive to shallow updates in 14.1.1, IIRC.

That being said, be aware that in `nuqs`, there will be two renders when setting a state: one with the internal state update, and one when the URL eventually updates (triggered by a reactive `useSearchParams`). This may or may not be of relevance for your application.

**Reply by MantasJa** [reply]
Thank you for your answer, my next.js version: 14.0.1

**Reply by franky47** [reply]
If you are stuck to this version, you could use the (now deprecated) `subscribeToQueryUpdates` API from `nuqs`. It will spy on both shallow and deep updates and will give you the URLSearchParams that have been applied to the URL.

But I'd rather recommend updating Next.js to latest, or at least to >=14.1.1 to have proper support for shallow updates.

**Reply by MantasJa** [reply]
You were right, after upgrade useEffect works with pahname/useSearchParams dependencies. Thank you
